Introduce Table of Content

ShaopengLin opened the pull request:
Fix #42 

Changes:
- Inject to pages a Table of Contents comprised of a nested hierarchy list of hyperlinks to <h> tags in the page.
- Add a button in the corner to toggle the table.
- No button and table are displayed if the table is empty.

kelson42 commented:
@ShaopengLin  The feature work well to my opinion, but we need to change/polish the way how the TOC is displayed, [see my comment in the issue](https://github.com/kiwix/kiwix-desktop/issues/42#issuecomment-2337479230).

On the top:
* Please put the TOC icon in the toolbar beside the random icon
* Create an entry in the menu "View > Table of Content"
* Create a keyboard shorcut to toggle the sidebar: ctrl+m

kelson42 commented:
> @kelson42 To clarify, what does the view entry do? I didn't see it from the mockup, but I assume you mean either a button to collapse the entire TOC, or it will do the same thing as the TOC icon. Like a show/hide button?

Not sure to fully understand you, but "yes", this is about toggling (hide/show) the TOC sidebar.

ShaopengLin commented:
@kelson42 Improved the TOC as much as possible requested in this comment https://github.com/kiwix/kiwix-desktop/pull/1201#issuecomment-2337533951.

kelson42 commented:
@ShaopengLin Sorry, I don't get it, shortcut does not display the TOC sidebar AND no icon in the toolbar 
![image](https://github.com/user-attachments/assets/1094c4d3-8665-4b78-aa78-dc887621c18d)

ShaopengLin commented:
@kelson42 I didn't find any issues on my end.

![Screenshot from 2024-09-10 08-11-03](https://github.com/user-attachments/assets/72220b59-33e8-4e9f-9cf3-a071dc94f2f7)

ShaopengLin commented:
@kelson42 

> Strategy should be clear if a section title is longer than the width: ellipsis en full title in the tooltip (when the mouse is over)
I will try my best on the ellipsis (Had some trouble with it on the CSS side). The tooltip should be doable.

> Secure look an feel is the same between bookmark and toc sidebar (The one of the TOC sidebar is better, it should be the reference)

Out of scope for this PR, I will open an issue. The reading list's UI refactoring is long overdue anyway...

> I believe it would be clearer if the full background is highlighted when the mouse goes over the items. For the moment we can not keep the selection I guess as it should then move dynamically when we scroll the article

Do you mean to highlight the items in the TOC when we hover them? Not sure what "full background" means here.

The dynamic scrolling might be too much for this PR... an indicator of localization has proven rather hard for me given the time constraint. I will open an issue afterward of this.

kelson42 commented:
> Do you mean to highlight the items in the TOC when we hover them? Not sure what "full background" means here.

I believe you understand properly. Item background highlighted from thel left to the right of the sidebar
 
> The dynamic scrolling might be too much for this PR... an indicator of localization has proven rather hard for me given the time constraint. I will open an issue afterward of this.

Indeed. Thx

ShaopengLin commented:
@kelson42 All issues mentioned are resolved:
1. TOC doesn't close on navigations and tab switches.
2. TOC and reading list does not co-exist.
3. grey out button appropriately
4. Added ellipsis and hover highlighting. Tool tips appears on the bottom left since they are links.

Once we finally agree on the UI I can refactor the code. The implementation is hard and due to time constraints, I didn't do this methodically. I will ping veloman when I finish by then.

kelson42 commented:
Lots of things work fine, but:
* TOC toolbar button is not always activated properly, here just after opening a ZIM file from the local library! I don't believe you have followed the same approach like the random button and seems to have build a new (broken) logic to decide if this button should be actived!
![image](https://github.com/user-attachments/assets/01b2f225-4694-4b7e-acf1-f96d8509b114)
* Please remove the "cursor" pointer, keep always the "pointer" mouse pointer.
![image](https://github.com/user-attachments/assets/1a32d49e-6127-46ed-8e10-1e44a83211b8)
* Item highlighting is OK, but there is a problem with right border not visible! highlighting left and right border should be the same
* ![image](https://github.com/user-attachments/assets/1a32d49e-6127-46ed-8e10-1e44a83211b8)
* Margin are still "wild". "Table of Content" (and then hide)  4 margin should be the same. Left margin of the first TOC item should be the same as "Table of Content" and aligned:
![image](https://github.com/user-attachments/assets/3cfbb553-186f-4bb0-b071-b72e07f018f5)
* Sidebar width is IMHO too big, should be the same as the bookmard sidebar
* By playing around with ctrl+M, ctrl+B and ctrl+R I still achieved to get the two bars displayed at the same time. The system to avoid duplicate sidebars should be robust and implemented differently
![image](https://github.com/user-attachments/assets/4980a5cf-014f-4fa3-9841-f85b00a8d956)

ShaopengLin commented:
@kelson42 Disable behavior now matches the random button. The rest of the UI request is fixed.

kelson42 commented:
@ShaopengLin Not all points: On the following screenshot you clear see:
* The width of the sidebar is even bigger than before and not the same as the bookmarking one
* Top margin over "highlighed" text is bigger than left
![image](https://github.com/user-attachments/assets/b51286c7-261f-427e-b172-301a562d0eea)

For the rest it looks good and when these two points are sorted out then we can probably start with code review

ShaopengLin commented:
@kelson42 Fix size done.

> Top margin over "highlighed" text is bigger than left
This isn't intentional. If you highlight any text, that's just how they are highlighted.

ShaopengLin commented:
@veloman-yunkan One question, am I allowed to introduce custom URL paths like "zim://kiwix-desktop-toc" ? Since we are using injected javascript, loading style sheet, icon, font, etc. would be much easier if we provide a request API path. It also is useful to pass information from the website to our application from some custom event.

kelson42 commented:
> @kelson42 Fix size done.
> 
> > Top margin over "highlighed" text is bigger than left
> 
> This isn't intentional. If you highlight any text, that's just how they are highlighted.

I hope with what follows, what is left to do is clearer:
* The red part should not exist
* the "hide" bottom should be the same as the "Table of content"
![image](https://github.com/user-attachments/assets/d33f4f84-7c0c-4592-8db6-e89fc8f412d7)

ShaopengLin commented:
@kelson42 disappearance should be fixed.

Unfortunately, the top and bottom padding around the text belong to line height, which the border cannot close in on.  Otherwise the distances are consistent.
![1](https://github.com/user-attachments/assets/82f10c4b-6d70-46ef-90da-f6f927851958)

veloman-yunkan commented:
> @veloman-yunkan The TOC has a hide button that hides it. Our application in Qt also has a toggle for TOC. However, our application has no way to know the user clicked the hide button inside TOC since its in the webpage. Thus, I am considering sending requests like "zim://kiwix-desktop-toc?close=true" or something like that to update state changes for our application.

Please let me research what solutions are possible and reply tomorrow.

kelson42 commented:
Don't abuse the zim:// scheme for anything else as accessing the ZIM file.

veloman-yunkan commented:
@ShaopengLin It looks like you can take advantage of [`QWebChannel`](https://doc.qt.io/qt-5/qwebchannel.html) (via [`QWebEnginePage::setWebChannel()`](https://doc.qt.io/qt-5/qwebenginepage.html#setWebChannel-1)).

If for some reason that or some other approach doesn't work and you have to resort to your hack of communicating via `UrlSchemeHandler` then you can use a new URL scheme for that (e.g. `kiwix-desktop://`).

ShaopengLin commented:
@veloman-yunkan Thanks! This is a lot more intuitive and works well.

@kelson42 I have addressed the issues of UI in this [comment](https://github.com/kiwix/kiwix-desktop/pull/1201#issuecomment-2346254895). If the UI is ready I can go on to the code refactoring.

kelson42 reviewed (approved):
From a user perspective LGTM
@veloman-yunkan Ready for the technical review

kelson42 commented:
@ShaopengLin Still a bug: If section title is shorten, then there is not hint displayed with the full title! 
![image](https://github.com/user-attachments/assets/fb0c0607-a27b-409e-abd9-c32d35f0cd41)

ShaopengLin commented:
@kelson42 I could not reproduce this on my own ZIM files. Could you direct me to where your example file is?
![Screenshot from 2024-09-16 07-46-16](https://github.com/user-attachments/assets/bc82879d-fccf-4902-b003-f46a4c90adcd)

kelson42 commented:
> @kelson42 I could not reproduce this on my own ZIM files. Could you direct me to where your example file is? ![Screenshot from 2024-09-16 07-46-16](https://private-user-images.githubusercontent.com/39098000/367741339-bc82879d-fccf-4902-b003-f46a4c90adcd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0ODg3MDcsIm5iZiI6MTcyNjQ4ODQwNywicGF0aCI6Ii8zOTA5ODAwMC8zNjc3NDEzMzktYmM4Mjg3OWQtZmNjZi00OTAyLWIwMDMtZjQ2YTRjOTBhZGNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDEyMDY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxYTdmZjU0ZWUzZjRkYWY4MzcxMjAzYzgyNmRiN2NmMWNiNjJjZmFmNmVhYzNmZTdlYmIwZDNmMGM5ZDI4NzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kF2alVcuQX3GhjIERfsGFfS_wTNJP-yp1DKqjoSnrUQ)

There is not tooltip on your screenshot either! I'm not talking about the tooltip at the bottom which display the targeted URL, I'm talking about a tooltip displyed at the section title displaying the title of the section (you have it already because you display it in the TOC).

kelson42 commented:
@ShaopengLin Any news abou the code cleaning you are running? It's important we start soon with code review.

ShaopengLin commented:
@kelson42 I have just finished the zim search one today. I will start on this shortly.

ShaopengLin commented:
@veloman-yunkan Code is ready for review.

veloman-yunkan commented:
I am about to review this PR (I know I should have done it long ago :man_facepalming: )

One question that is probably too late to ask but I'd like to know the answer nevertheless. Given that the Bookmark and TOC sidebars must be mutually exclusive, couldn't the TOC sidebar be implemented akin to the Bookmark sidebar (i.e. without modifying the webpage)?

Also will you please rebase the PR?

ShaopengLin commented:
@veloman-yunkan Actually this rings a bell now as I can now pass information from Javascript to Qt with QWebChannel. I quite often get stuck in the same mindset after working on something for very long...

I will get back to you to see if I can pass the table and links back to a QTreeView/Widget in a reasonable manner.

veloman-yunkan commented:
If you are going to provide an alternative implementation please do that in a separate PR - don't change this one.

ShaopengLin commented:
@veloman-yunkan I will try to get the alternative version up (or prove it doesn't work, though it seems to be promising from my rough tests) in the coming days. In a rush for paper submission so it will likely be there around the weekends.

kelson42 commented:
@ShaopengLin Unfortunately a new rebase will be necessary!

kelson42 commented:
Superseeded by #1237
